### PR TITLE
Stasis bed TLC: Emag effects, toggleable and upgradable blood filtering

### PR DESCRIPTION
**Pull request by Metekillot**
## About The Pull Request

This is to tackle a couple of things. The stasis bed is one of the fundamental aspects of medbay, but there are no options to sabotage it besides turning it off. This PR adds a little something to do that; emagging a stasis bed will make anyone buckled into it incapacitated and forced to speak at a whisper so they can only call for help from people nearby.

Emagging also has an effect on the second part of this PR: blood filtering for stasis beds. It works as follows:

You are able to toggle a blood filtering option on for stasis beds. This causes anyone buckled into it (or already buckled in it) to take 5 brute damage and spatter a little blood. If it's emagged, you take 20 brute damage from this instead. This only happens upon initial buckling or initial activation. Sly subverted silicons or malicious doctors can repeatedly toggle this setting, though the cooldown means that it's usually faster to just hit you in the head with a saw a lot.

From there, depending on your upgrade tier, the stasis bed will either:
Gradually filter reagents from your blood, or if at higher tiers, it will avoid filtering out medical reagents in your blood.

The amount of reagents the bed filters when enabled in this mode is determined by its filtering_efficiency, which is itself determined first by its servo tier, then modified by its capacitor tier. This maxes out at 8 units of each reagent filtered out at the highest level.

But if it's emagged? Something else happens: While you're incapacitated and forced to a whisper, the stasis bed will ONLY filter out medicine from your blood, and it will also gradually drain your blood. The rate of this is determined by the tier of the parts, as well, so it pays for any emaggers to wait until later in the round to emag the fully upgraded stasis beds.

If you try to get your blood filtered as a jellyperson, it will drain your blood as if it were emagged, but otherwise drain toxic reagent out of you, if you wanted to commit a convoluted suicide for some reason... or if a bored antagonist bucklecuffs you to it.
## Why It's Good For The Game

We have a few ways to filter toxins from people right now. All of them take forever and are a bit of a pain in the ass. This is an option for lazy doctors to filter toxins, the same way that the cryo tube is an option for lazy doctors to treat patients. Right now we do have the potassium+water jank, but that's glitchy tribal knowledge and probably not an intended medical treatment.

The emagging is to offer another sly avenue for antagonists to fuck with medbay, or to set up a ripper clinic in maintenance full of emagged stasis beds just waiting for some helpless victims.
## Changelog

Stasis beds now have an upgradable blood filtering toggle, and are now able to be emagged.
:cl: Bisar
add: Added emag interactions to stasis beds.
add: Added optional and upgradable function to sleeper beds for filtering blood. Jellypeople are advised against using it.
/:cl:

### CODE REVIEW

File: code/game/machinery/stasis.dm [hunk 1]
```diff
@@ -50,13 +80,51 @@
...
+/// Toggle the filtering function on or off. This makes a bunch of scary-ass needles poke out of the bed.
+/// In exchange for having a passive function on hand to filter people's blood, it causes a little bit of damage.
+/// This damage only occurs upon initial activation, either through filtering starting by you being buckled or
+/// someone activating filtering on the bed you're buckled in already. Malicious doctors or silicons could toggle
+/// this over and over while you're helplessly buckled to the bed, though the effect of you being mutilated with
+/// needles will be obvious to onlookers.
+/obj/machinery/stasis/CtrlClick(mob/user)
+	if(!can_interact(user))
+		return
+	. = ..()
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:92]:
```suggestion
	. = ..()
	if(!can_interact(user))
		return
```
Though `/atom/proc/CtrlClick()` already calls `can_interact` so this is wasteful but it doesn't matter that much

File: code/game/machinery/stasis.dm [hunk 2]
```diff
@@ -50,13 +80,51 @@
...
+	if(world.time < can_toggle)
+		return CLICK_ACTION_BLOCKING
+	filtering_enabled = !filtering_enabled
+	can_toggle = world.time + STASIS_TOGGLE_COOLDOWN
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:96]: Use the cooldown defines. `COOLDOWN_START` `COOLDOWN_FINISHED` etc

File: code/game/machinery/stasis.dm [hunk 3]
```diff
@@ -50,13 +80,51 @@
...
+	if(filtering_enabled)
+		var/sound/toggle_sound = sound('sound/items/unsheath.ogg')
+		toggle_sound.pitch = 2
+		playsound(src, toggle_sound, 60, TRUE,)
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:100]:
```suggestion
		playsound(src, toggle_sound, 60, TRUE)
```

File: code/game/machinery/stasis.dm [hunk 4]
```diff
@@ -50,13 +80,51 @@
...
+	else
+		var/sound/toggle_sound = sound('sound/items/sheath.ogg')
+		toggle_sound.pitch = 2
+		playsound(src, toggle_sound, 60, TRUE,)
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:104]:
```suggestion
		playsound(src, toggle_sound, 60, TRUE)
```

File: code/game/machinery/stasis.dm [hunk 5]
```diff
@@ -50,13 +80,51 @@
...
+	user.visible_message(span_notice("\The [src] filtering function [filtering_enabled ? "powers on. Thousands of micro-needles jut up from the mattress" : "shuts down. The micro-needles shunt back into the mattress invisibly"]."), \
+				span_notice("You [filtering_enabled ? "power on" : "shut down"] \the [src] filtering function. [filtering_enabled ? "Thousands of micro-needles jut up from the mattress" : "The micro-needles collapse back into the mattress, invisible"]."), \
+				span_hear("You hear [filtering_enabled ? " the unnerving sound of sharpened metal sliding against metal" : " a clunk of something metal"]."))
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:107]:
```suggestion
	user.visible_message(
		span_notice("[src] filtering function [filtering_enabled ? "powers on. Thousands of micro-needles jut up from the mattress" : "shuts down. The micro-needles shunt back into the mattress invisibly"]."),
		span_notice("You [filtering_enabled ? "power on" : "shut down"] [src] filtering function. [filtering_enabled ? "Thousands of micro-needles jut up from the mattress" : "The micro-needles collapse back into the mattress, invisible"]."), 
		span_hear("You hear [filtering_enabled ? " the unnerving sound of sharpened metal sliding against metal" : " a clunk of something metal"]."),
	)
```
`\The` is unnecessary in `\The [src]` also we can multiline this in a way easier to read

File: code/game/machinery/stasis.dm [hunk 6]
```diff
@@ -107,19 +175,81 @@
...
+		target.add_traits(list(TRAIT_INCAPACITATED, TRAIT_SOFTSPOKEN), STASIS_MACHINE_EFFECT)
 	target.extinguish_mob()
-	update_use_power(ACTIVE_POWER_USE)
+	/// Double power usage if we have filtering enabled.
+	update_use_power(ACTIVE_POWER_USE * (filtering_enabled ? 2 : 1))
+
+/obj/machinery/stasis/proc/filter_jab(mob/living/target, mob/user)
+	target.visible_message(span_danger("The thousands of micro-needles in the [src]'s mattress slide into [target]'s flesh."), \
+				span_danger("You feel a sharp jabbing sensation as the needles of the [src] pierce your skin!"), \
+				span_hear("You hear a wet squelching sound."))
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:192]:
```suggestion
	target.visible_message(
		span_danger("The thousands of micro-needles in [src]'s mattress slide into [target]'s flesh."), 
		span_danger("You feel a sharp jabbing sensation as the needles of [src] pierce your skin!"),
		span_hear("You hear a wet squelching sound."),
	)
```

File: code/game/machinery/stasis.dm [hunk 7]
```diff
@@ -107,19 +175,81 @@
...
+		onlooker_message = span_notice("[src] filters [target]'s blood, emitting an ominous whirring noise.")
+		target_message = span_danger("[src] filters your blood-- SOMETHING IS WRONG! You feel emptier!")
+		target.blood_volume = max(target.blood_volume - (filtering_efficiency * 3), 0)
+		if(prob(15))
+			target.emote("scream")
+	else
+		onlooker_message = span_notice("[src] filters [target]'s blood, humming softly.")
+		target_message = span_notice("You feel an odd, floaty sensation as [src] filters your blood.")
+	target.visible_message(onlooker_message, target_message, span_hear("You hear a wet, squelching, slurping sort of noise."), vision_distance = 1)
+	for(var/datum/reagent/to_remove in target.reagents.reagent_list)
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:220]:
```suggestion
	for(var/datum/reagent/to_remove as anything in target.reagents.reagent_list)
```

File: code/game/machinery/stasis.dm [hunk 8]
```diff
@@ -107,19 +175,81 @@
...
+		if(is_emagged)
+			if(istype(to_remove, /datum/reagent/medicine/))
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:222]:
```suggestion
			if(istype(to_remove, /datum/reagent/medicine))
```

File: code/game/machinery/stasis.dm [hunk 9]
```diff
@@ -107,19 +175,81 @@
...
+				target.reagents.remove_reagent(to_remove.type, filtering_efficiency)
+				/// If we're emagged, we're going to be a bit more "aggressive" with filtering.
+				/// And a bit more forgetful about actually filtering out the right things.
+				continue
+		if(filtering_efficiency < 3)
+			target.reagents.remove_reagent(to_remove.type, filtering_efficiency)
+		else
+			if(istype(to_remove, /datum/reagent/medicine/))
```
Comment (MrMelbert) [inline code/game/machinery/stasis.dm:230]:
```suggestion
			if(istype(to_remove, /datum/reagent/medicine))
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]
So this completely powercreeps the blood filter surgical tool?

**Comment by Metekillot** [comment]
> So this completely powercreeps the blood filter surgical tool?

I had wanted to do something with the blood filter as well to make it improvable, but it was outside the scope of this particular PR and a broader intent to address the jank we have for handling chems and overdoses.

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]
I'm not really open to adding functionality to stasis beds.
They are not intended to be used for surgeries and should do the bare minimum of stasis.
This is not a QOL this is a feature.

**Comment by Iamgoofball** [comment]
Stasis beds need the ability for the person on them to press a clicker to request attention from medical, even if unconscious. They also should hook to surgery computers but without allowing advanced surgery access.

**Comment by Metekillot** [comment]
> I'm not really open to adding functionality to stasis beds. They are not intended to be used for surgeries and should do the bare minimum of stasis. This is not a QOL this is a feature.

I am not sure why you are telling me it's not a QOL. I didn't mark it as a QOL.

**Comment by mc-oofert** [comment]
so you're telling me I can bait people to move on top of a stasis bed and I can immediately incapacitate them if emagged including if I build one in maints

**Comment by Metekillot** [comment]
Well I believe they glow in the dark though I didn't check that when I was making this. But yes. If you manage to trick someone into standing on top of a stasis bed for long enough for you to buckle them after you have emagged it, you would have subdued this hypothetical person.

**Comment by mc-oofert** [comment]
buckling people on the same tile is instant

**Comment by Metekillot** [comment]
Yes, it is.

**Comment by ZephyrTFA** [comment]
I'm not interested in having stasis beds be more than stasis beds

**Pull request closed without merging**